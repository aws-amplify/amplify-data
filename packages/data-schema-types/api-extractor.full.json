{
  "extends": "../../api-extractor.full.base.json",
  "mainEntryPointFilePath": "<projectFolder>/dist/esm/index.internal.d.ts"
}
