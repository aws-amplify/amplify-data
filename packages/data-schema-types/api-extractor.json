{
  "extends": "../../api-extractor.base.json",
  "mainEntryPointFilePath": "<projectFolder>/dist/esm/index.d.ts"
}
