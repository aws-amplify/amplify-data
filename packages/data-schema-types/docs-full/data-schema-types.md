<!-- Do not edit this file. It is automatically generated by API Documenter. -->

[Home](./index.md) &gt; [@aws-amplify/data-schema-types](./data-schema-types.md)

## data-schema-types package

## Interfaces

<table><thead><tr><th>

Interface


</th><th>

Description


</th></tr></thead>
<tbody><tr><td>

[DerivedApiDefinition](./data-schema-types.derivedapidefinition.md)


</td><td>


</td></tr>
<tr><td>

[GraphQLFormattedError](./data-schema-types.graphqlformattederror.md)


</td><td>

See: https://spec.graphql.org/draft/\#sec-Errors


</td></tr>
<tr><td>

[SourceLocation](./data-schema-types.sourcelocation.md)


</td><td>

Represents a location in a Source.


</td></tr>
</tbody></table>

## Variables

<table><thead><tr><th>

Variable


</th><th>

Description


</th></tr></thead>
<tbody><tr><td>

[\_\_modelMeta\_\_](./data-schema-types.__modelmeta__.md)


</td><td>


</td></tr>
</tbody></table>

## Type Aliases

<table><thead><tr><th>

Type Alias


</th><th>

Description


</th></tr></thead>
<tbody><tr><td>

[AuthMode](./data-schema-types.authmode.md)


</td><td>


</td></tr>
<tr><td>

[BackendSecret](./data-schema-types.backendsecret.md)


</td><td>

Importing the full objects from @<!-- -->aws-amplify/plugin-types more than doubles dev env runtime. This type replacement will contain the content for config without the negative side-effects. We may need to re-approach if customers interact with these programmatically to avoid forcing narrowing.


</td></tr>
<tr><td>

[BooleanFilters](./data-schema-types.booleanfilters.md)


</td><td>


</td></tr>
<tr><td>

[ContextType](./data-schema-types.contexttype.md)


</td><td>


</td></tr>
<tr><td>

[CreateModelInput](./data-schema-types.createmodelinput.md)


</td><td>

All identifiers and fields used to create a model


</td></tr>
<tr><td>

[CustomHeaders](./data-schema-types.customheaders.md)


</td><td>

Custom headers that can be passed either to the client or to individual model operations, either as a static object or a function that returns a promise.


</td></tr>
<tr><td>

[CustomMutations](./data-schema-types.custommutations.md)


</td><td>


</td></tr>
<tr><td>

[CustomOperationFnParams](./data-schema-types.customoperationfnparams.md)


</td><td>

Generates Custom Operations function params based on whether .arguments() were specified in the schema builder


</td></tr>
<tr><td>

[CustomOperationMethodOptions](./data-schema-types.customoperationmethodoptions.md)


</td><td>


</td></tr>
<tr><td>

[CustomOperations](./data-schema-types.customoperations.md)


</td><td>


</td></tr>
<tr><td>

[CustomQueries](./data-schema-types.customqueries.md)


</td><td>


</td></tr>
<tr><td>

[CustomSelectionSetReturnValue](./data-schema-types.customselectionsetreturnvalue.md)


</td><td>

Generates flattened, readonly return type using specified custom sel. set


</td></tr>
<tr><td>

[CustomSqlDataSourceStrategy](./data-schema-types.customsqldatasourcestrategy.md)


</td><td>


</td></tr>
<tr><td>

[CustomSubscriptions](./data-schema-types.customsubscriptions.md)


</td><td>


</td></tr>
<tr><td>

[DataSourceConfiguration](./data-schema-types.datasourceconfiguration.md)


</td><td>


</td></tr>
<tr><td>

[DatasourceEngine](./data-schema-types.datasourceengine.md)


</td><td>


</td></tr>
<tr><td>

[Debug](./data-schema-types.debug.md)


</td><td>


</td></tr>
<tr><td>

[DeepPickFromPath](./data-schema-types.deeppickfrompath.md)


</td><td>

Picks object properties that match provided dot-separated Path


</td></tr>
<tr><td>

[DeepReadOnlyObject](./data-schema-types.deepreadonlyobject.md)


</td><td>


</td></tr>
<tr><td>

[DefineFunction](./data-schema-types.definefunction.md)


</td><td>


</td></tr>
<tr><td>

[DerivedCombinedSchema](./data-schema-types.derivedcombinedschema.md)


</td><td>


</td></tr>
<tr><td>

[DerivedModelSchema](./data-schema-types.derivedmodelschema.md)


</td><td>


</td></tr>
<tr><td>

[EnumTypes](./data-schema-types.enumtypes.md)


</td><td>

The utility type that is used to infer the type (interface) of the generated `client.enums` property.


</td></tr>
<tr><td>

[Equal](./data-schema-types.equal.md)


</td><td>


</td></tr>
<tr><td>

[ExcludeEmpty](./data-schema-types.excludeempty.md)


</td><td>


</td></tr>
<tr><td>

[Expect](./data-schema-types.expect.md)


</td><td>


</td></tr>
<tr><td>

[ExpectFalse](./data-schema-types.expectfalse.md)


</td><td>


</td></tr>
<tr><td>

[ExpectTrue](./data-schema-types.expecttrue.md)


</td><td>


</td></tr>
<tr><td>

[ExtractModelMeta](./data-schema-types.extractmodelmeta.md)


</td><td>


</td></tr>
<tr><td>

[FunctionSchemaAccess](./data-schema-types.functionschemaaccess.md)


</td><td>


</td></tr>
<tr><td>

[HasKey](./data-schema-types.haskey.md)


</td><td>


</td></tr>
<tr><td>

[IfEquals](./data-schema-types.ifequals.md)


</td><td>


</td></tr>
<tr><td>

[IndexQueryMethodsFromIR](./data-schema-types.indexquerymethodsfromir.md)


</td><td>


</td></tr>
<tr><td>

[IndexQueryMethodSignature](./data-schema-types.indexquerymethodsignature.md)


</td><td>


</td></tr>
<tr><td>

[IsAny](./data-schema-types.isany.md)


</td><td>


</td></tr>
<tr><td>

[IsEmptyStringOrNever](./data-schema-types.isemptystringornever.md)


</td><td>


</td></tr>
<tr><td>

[IsFalse](./data-schema-types.isfalse.md)


</td><td>


</td></tr>
<tr><td>

[IsTrue](./data-schema-types.istrue.md)


</td><td>


</td></tr>
<tr><td>

[JsResolver](./data-schema-types.jsresolver.md)


</td><td>


</td></tr>
<tr><td>

[JsResolverEntry](./data-schema-types.jsresolverentry.md)


</td><td>


</td></tr>
<tr><td>

[LambdaFunctionDefinition](./data-schema-types.lambdafunctiondefinition.md)


</td><td>


</td></tr>
<tr><td>

[LazyLoader](./data-schema-types.lazyloader.md)


</td><td>


</td></tr>
<tr><td>

[ListReturnValue](./data-schema-types.listreturnvalue.md)


</td><td>


</td></tr>
<tr><td>

[LogicalFilters](./data-schema-types.logicalfilters.md)


</td><td>


</td></tr>
<tr><td>

[Model](./data-schema-types.model.md)


</td><td>


</td></tr>
<tr><td>

[ModelFilter](./data-schema-types.modelfilter.md)


</td><td>


</td></tr>
<tr><td>

[ModelIdentifier](./data-schema-types.modelidentifier.md)


</td><td>


</td></tr>
<tr><td>

[ModelMetaShape](./data-schema-types.modelmetashape.md)


</td><td>


</td></tr>
<tr><td>

[ModelPath](./data-schema-types.modelpath.md)


</td><td>

Generates custom selection set type with up to 6 levels of nested fields


</td></tr>
<tr><td>

[ModelSortDirection](./data-schema-types.modelsortdirection.md)


</td><td>


</td></tr>
<tr><td>

[ModelTypes](./data-schema-types.modeltypes.md)


</td><td>


</td></tr>
<tr><td>

[ModelTypesClient](./data-schema-types.modeltypesclient.md)


</td><td>


</td></tr>
<tr><td>

[ModelTypesSSRCookies](./data-schema-types.modeltypesssrcookies.md)


</td><td>


</td></tr>
<tr><td>

[ModelTypesSSRRequest](./data-schema-types.modeltypesssrrequest.md)


</td><td>


</td></tr>
<tr><td>

[MutationInput](./data-schema-types.mutationinput.md)


</td><td>

All required fields and relational fields, exclude readonly fields


</td></tr>
<tr><td>

[NonRelationalFields](./data-schema-types.nonrelationalfields.md)


</td><td>

Currently this omits any object-type fields. Update this when we add custom types/enums.


</td></tr>
<tr><td>

[NotAny](./data-schema-types.notany.md)


</td><td>


</td></tr>
<tr><td>

[NotEqual](./data-schema-types.notequal.md)


</td><td>


</td></tr>
<tr><td>

[NumericFilter](./data-schema-types.numericfilter.md)


</td><td>


</td></tr>
<tr><td>

[ObjectIsNonEmpty](./data-schema-types.objectisnonempty.md)


</td><td>


</td></tr>
<tr><td>

[ObservedReturnValue](./data-schema-types.observedreturnvalue.md)


</td><td>


</td></tr>
<tr><td>

[ObserveQueryReturnValue](./data-schema-types.observequeryreturnvalue.md)


</td><td>


</td></tr>
<tr><td>

[PathEntry](./data-schema-types.pathentry.md)


</td><td>

references IAmplifyGraphqlDefinition from: https://github.com/aws-amplify/amplify-category-api/blob/4c0ea253a0bae51f775383929ba4748593185bc1/packages/amplify-graphql-api-construct/src/types.ts\#L491-L503

function slots is any'd for now. Will add actual type when we add support for this feature


</td></tr>
<tr><td>

[Prettify](./data-schema-types.prettify.md)


</td><td>


</td></tr>
<tr><td>

[RequestOptions](./data-schema-types.requestoptions.md)


</td><td>

Request options that are passed to custom header functions. `method` and `headers` are not included in custom header functions passed to subscriptions.


</td></tr>
<tr><td>

[ResolvedModel](./data-schema-types.resolvedmodel.md)


</td><td>

Flattens model instance type and unwraps async functions into resolved GraphQL shape

This type is used for generating the base shape for custom selection set input and its return value Uses same pattern as above to limit recursion depth to maximum usable for selection set.


</td></tr>
<tr><td>

[RestoreArrays](./data-schema-types.restorearrays.md)


</td><td>

This mapped type traverses the SelectionSetReturnValue result and the original FlatModel, restoring array types that were flattened in DeepPickFromPath


</td></tr>
<tr><td>

[ReturnValue](./data-schema-types.returnvalue.md)


</td><td>

Selection set-aware CRUDL operation return value type


</td></tr>
<tr><td>

[SchemaConfiguration](./data-schema-types.schemaconfiguration.md)


</td><td>


</td></tr>
<tr><td>

[SecondaryIndexIrShape](./data-schema-types.secondaryindexirshape.md)


</td><td>

SecondaryIndex index types and query methods


</td></tr>
<tr><td>

[SelectionSet](./data-schema-types.selectionset.md)


</td><td>


</td></tr>
<tr><td>

[SetTypeSubArg](./data-schema-types.settypesubarg.md)


</td><td>

Replaces the value of a key in a complex generic type param


</td></tr>
<tr><td>

[SingularReturnValue](./data-schema-types.singularreturnvalue.md)


</td><td>


</td></tr>
<tr><td>

[SizeFilter](./data-schema-types.sizefilter.md)


</td><td>

Filter options that can be used on fields where size checks are supported.


</td></tr>
<tr><td>

[SqlStatementFolderEntry](./data-schema-types.sqlstatementfolderentry.md)


</td><td>


</td></tr>
<tr><td>

[StringFilter](./data-schema-types.stringfilter.md)


</td><td>

Filters options that can be used on string-like fields.


</td></tr>
<tr><td>

[SubnetAZ](./data-schema-types.subnetaz.md)


</td><td>


</td></tr>
<tr><td>

[UnionToIntersection](./data-schema-types.uniontointersection.md)


</td><td>


</td></tr>
<tr><td>

[UnwrapArray](./data-schema-types.unwraparray.md)


</td><td>


</td></tr>
<tr><td>

[VpcConfig](./data-schema-types.vpcconfig.md)


</td><td>


</td></tr>
<tr><td>

[WritableKeys](./data-schema-types.writablekeys.md)


</td><td>


</td></tr>
</tbody></table>
