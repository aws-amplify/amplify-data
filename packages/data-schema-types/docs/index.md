<!-- Do not edit this file. It is automatically generated by API Documenter. -->

[Home](./index.md)

## API Reference

## Packages

|  Package | Description |
|  --- | --- |
|  [@aws-amplify/data-schema-types](./data-schema-types.md) |  |

