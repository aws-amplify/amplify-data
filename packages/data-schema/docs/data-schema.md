<!-- Do not edit this file. It is automatically generated by API Documenter. -->

[Home](./index.md) &gt; [@aws-amplify/data-schema](./data-schema.md)

## data-schema package

## Enumerations

<table><thead><tr><th>

Enumeration


</th><th>

Description


</th></tr></thead>
<tbody><tr><td>

[ModelRelationshipTypes](./data-schema.modelrelationshiptypes.md)


</td><td>

Model relationship types


</td></tr>
</tbody></table>

## Interfaces

<table><thead><tr><th>

Interface


</th><th>

Description


</th></tr></thead>
<tbody><tr><td>

[EnumType](./data-schema.enumtype.md)


</td><td>

Enum type definition content


</td></tr>
</tbody></table>

## Namespaces

<table><thead><tr><th>

Namespace


</th><th>

Description


</th></tr></thead>
<tbody><tr><td>

[a](./data-schema.a.md)


</td><td>


</td></tr>
</tbody></table>

## Type Aliases

<table><thead><tr><th>

Type Alias


</th><th>

Description


</th></tr></thead>
<tbody><tr><td>

[Authorization](./data-schema.authorization.md)


</td><td>

Container for authorization schema definition content.


</td></tr>
<tr><td>

[ClientSchema](./data-schema.clientschema.md)


</td><td>


</td></tr>
<tr><td>

[CombinedModelSchema](./data-schema.combinedmodelschema.md)


</td><td>

CombinedModel schema definition interface


</td></tr>
<tr><td>

[CustomOperation](./data-schema.customoperation.md)


</td><td>

Custom operation definition interface


</td></tr>
<tr><td>

[CustomType](./data-schema.customtype.md)


</td><td>

Custom type container


</td></tr>
<tr><td>

[Json](./data-schema.json.md)


</td><td>

A precise, recursive Json type blows the type calculation stack without installing explicit `Json extends T ? short-circuit : ...` type checks all over the place. We may take that on later. But, this is a good-enough approximation for now.


</td></tr>
<tr><td>

[ModelDefaultIdentifier](./data-schema.modeldefaultidentifier.md)


</td><td>

Model default identifier


</td></tr>
<tr><td>

[ModelField](./data-schema.modelfield.md)


</td><td>

Public API for the chainable builder methods exposed by Model Field. The type is narrowing e.g., after calling .array() it will be omitted from intellisense suggestions


</td></tr>
<tr><td>

[ModelRelationshipField](./data-schema.modelrelationshipfield.md)


</td><td>

Model relationship field definition interface


</td></tr>
<tr><td>

[ModelRelationshipTypeArgFactory](./data-schema.modelrelationshiptypeargfactory.md)


</td><td>

Model relationship type definition content


</td></tr>
<tr><td>

[ModelSchema](./data-schema.modelschema.md)


</td><td>

Model schema definition interface


</td></tr>
<tr><td>

[ModelType](./data-schema.modeltype.md)


</td><td>

Model type definition interface


</td></tr>
<tr><td>

[Nullable](./data-schema.nullable.md)


</td><td>

Field type arg mutators


</td></tr>
<tr><td>

[RDSModelSchema](./data-schema.rdsmodelschema.md)


</td><td>

RDSModel schema definition interface


</td></tr>
<tr><td>

[RefType](./data-schema.reftype.md)


</td><td>

Reference type definition interface


</td></tr>
<tr><td>

[ResourceAuthorization](./data-schema.resourceauthorization.md)


</td><td>

Container for representing the authorization of function resources


</td></tr>
</tbody></table>
