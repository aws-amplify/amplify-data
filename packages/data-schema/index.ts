import * as a from './src';
import { ClientSchema } from './src/ClientSchema';

export { a };

export type { ClientSchema };

export * from './src/runtime/client';
export {
  addSchemaToClient,
  addSchemaToClientWithInstance,
} from './src/runtime';
