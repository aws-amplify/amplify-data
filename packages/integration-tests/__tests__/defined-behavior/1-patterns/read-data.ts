import { a, ClientSchema } from '@aws-amplify/data-schema';
import { Amplify } from 'aws-amplify';
import { type SelectionSet } from '@aws-amplify/data-schema-types';
import {
  buildAmplifyConfig,
  mockedGenerateClient,
  optionsAndHeaders,
  useState,
} from '../../utils';

describe('Read application data', () => {
  // https://docs.amplify.aws/gen2/build-a-backend/data/query-data/

  const sampleTodo = {
    __typename: 'Todo',
    id: 'some-id',
    content: 'some content',
    description: 'something something',
    owner: 'some-body',
    priority: 2,
    done: false,
    updatedAt: '2024-03-01T19:05:44.536Z',
    createdAt: '2024-03-01T18:05:44.536Z',
  };

  // data/resource.ts
  const schema = a.schema({
    Todo: a
      .model({
        content: a.string(),
        description: a.string(),
        done: a.boolean(),
        priority: a.integer(),
      })
      .authorization((allow) => [
        allow.owner(),
        allow.publicApiKey().to(['read']),
      ]),
    Inventory: a
      .model({
        productId: a.id().required(),
        warehouseId: a.id().required(),
        description: a.string(),
      })
      .identifier(['productId', 'warehouseId'])
      .authorization((allow) => [
        allow.owner(),
        allow.publicApiKey().to(['read']),
      ]),
  });
  type Schema = ClientSchema<typeof schema>;

  afterEach(() => {
    jest.clearAllMocks();
  });

  test('List and get your data', async () => {
    // #region mocking
    const sampleInventory = {
      __typeName: 'Inventory',
      productId: 'abc',
      warehouseId: 'xyz',
      description: 'a widget',
      updatedAt: '2024-03-01T19:05:44.536Z',
      createdAt: '2024-03-01T18:05:44.536Z',
    };

    const { spy, generateClient } = mockedGenerateClient([
      {
        data: {
          listTodos: {
            items: [sampleTodo],
          },
        },
      },
      {
        data: {
          listInventories: [sampleInventory],
        },
      },
      {
        data: {
          getInventory: sampleInventory,
        },
      },
    ]);

    // simulated amplifyconfiguration.json
    const config = await buildAmplifyConfig(schema);
    // #endregion mocking

    // #region covers d802e473c55f60ff
    Amplify.configure(config);

    const client = generateClient<Schema>();

    // list all items
    const { data: todos, errors: listTodoErrors } =
      await client.models.Todo.list();

    // list -> index query
    const { data: inventories, errors: listInventoryErrors } =
      await client.models.Inventory.list({
        productId: '...',
        warehouseId: {
          beginsWith: '...',
        },
      });

    // get a specific item
    const { data: inventory, errors: getInventoryErrors } =
      await client.models.Inventory.get({
        productId: '...',
        warehouseId: '...',
      });
    // #endregion docs code

    // #region assertions
    expect(optionsAndHeaders(spy)).toMatchSnapshot();
    expect(listTodoErrors).toBeUndefined();
    expect(listInventoryErrors).toBeUndefined();
    expect(getInventoryErrors).toBeUndefined();
    expect(todos).toEqual([sampleTodo]);
    expect(inventories).toEqual([sampleInventory]);
    expect(inventory).toEqual(sampleInventory);
    // #endregion assertions
  });

  test('Filter list queries', async () => {
    // #region mocking
    const { spy, generateClient } = mockedGenerateClient([
      {
        data: {
          listTodos: {
            items: [sampleTodo],
          },
        },
      },
    ]);

    // simulated amplifyconfiguration.json
    const config = await buildAmplifyConfig(schema);
    // #endregion mocking

    // #region covers b7a13dcc42cc0f2b
    Amplify.configure(config);

    const client = generateClient<Schema>();

    const { data: todos, errors } = await client.models.Todo.list({
      filter: {
        content: {
          beginsWith: 'hello',
        },
      },
    });
    // #endregion

    // #region assertions
    expect(optionsAndHeaders(spy)).toMatchSnapshot();
    expect(errors).toBeUndefined();
    expect(todos).toEqual([sampleTodo]);
    // #endregion assertions
  });

  test('Compound filters', async () => {
    // #region mocking
    const { spy, generateClient } = mockedGenerateClient([
      {
        data: {
          listTodos: {
            items: [sampleTodo],
          },
        },
      },
    ]);

    // simulated amplifyconfiguration.json
    const config = await buildAmplifyConfig(schema);
    // #endregion mocking

    // #region covers 805e7a16e0b5b803
    Amplify.configure(config);

    const client = generateClient<Schema>();

    const { data: todos, errors } = await client.models.Todo.list({
      filter: {
        or: [
          {
            priority: { eq: 1 },
          },
          {
            priority: { eq: 2 },
          },
        ],
      },
    });
    // #endregion

    // #region assertions
    expect(optionsAndHeaders(spy)).toMatchSnapshot();
    expect(errors).toBeUndefined();
    expect(todos).toEqual([sampleTodo]);
    // #endregion assertions
  });

  test('Paginate list queries', async () => {
    // #region mocking
    const { spy, generateClient } = mockedGenerateClient([
      {
        data: {
          listTodos: {
            items: [sampleTodo],
            nextToken: 'some token',
          },
        },
      },
    ]);

    // simulated amplifyconfiguration.json
    const config = await buildAmplifyConfig(schema);
    // #endregion mocking

    // #region covers 6cac5a92233f7f71, af99c7c3ce818ba9
    // (`af99c7c3ce818ba9` demonstrates pagination in UI)
    Amplify.configure(config);

    const client = generateClient<Schema>();

    const {
      data: todos,
      nextToken, // Repeat this API call with the nextToken until the returned nextToken is `null`
      errors,
    } = await client.models.Todo.list({
      limit: 100, // default value is 100
      nextToken: 'eyJ2ZXJzaW9uejE1a2...', // previous nextToken
    });

    // front-end sample code skipped for unit test. we're not mocking full react DX for now.

    // #endregion

    // #region assertions
    expect(optionsAndHeaders(spy)).toMatchSnapshot();
    expect(errors).toBeUndefined();
    expect(todos).toEqual([sampleTodo]);
    // #endregion assertions
  });

  test('Fetch only the data you need with custom selection set', async () => {
    // #region mocking
    // This example uses a more complicated schema to demonstrate CSS
    const schema = a
      .schema({
        Blog: a.model({
          author: a.hasOne('Author', 'blogId'),
          content: a.hasMany('Content', 'blogId'),
          publication: a.hasOne('Publication', 'blogId'),
        }),
        Content: a.model({
          title: a.string(),
          description: a.string(),
          blogId: a.id(),
          blog: a.belongsTo('Blog', 'blogId'),
        }),
        Author: a.model({
          email: a.string(),
          blogId: a.id(),
          blog: a.belongsTo('Blog', 'blogId'),
        }),
        Publication: a.model({
          company: a.hasOne('Company', 'publicationId'),
          blogId: a.id(),
          blog: a.belongsTo('Blog', 'blogId'),
        }),
        Company: a.model({
          name: a.string().required(),
          publicationId: a.id(),
          publication: a.belongsTo('Publication', 'publicationId'),
          location: a.hasOne('Location', 'companyId'),
        }),
        Location: a.model({
          city: a.string(),
          state: a.string(),
          companyId: a.id(),
          company: a.belongsTo('Company', 'companyId'),
        }),
      })
      .authorization((allow) => allow.publicApiKey());
    type Schema = ClientSchema<typeof schema>;

    const blogWithSelectionSet = {
      __typeName: 'Blog',
      author: {
        email: 'someone@example.com',
      },
      publication: {
        company: {
          location: {
            city: 'wherever',
          },
        },
      },
      content: {
        items: [
          {
            title: 'some title',
            description: 'some description',
            id: 'some-id',
            updatedAt: '2024-03-01T19:05:44.536Z',
            createdAt: '2024-03-01T18:05:44.536Z',
          },
        ],
      },
    };

    // Final result will be flattened and not quite quite equal to the graphql response.
    const blogWithSelectionSetResult = {
      __typeName: 'Blog',
      author: {
        email: 'someone@example.com',
      },
      publication: {
        company: {
          location: {
            city: 'wherever',
          },
        },
      },
      content: [
        {
          title: 'some title',
          description: 'some description',
          id: 'some-id',
          updatedAt: '2024-03-01T19:05:44.536Z',
          createdAt: '2024-03-01T18:05:44.536Z',
        },
      ],
    };

    const { spy, generateClient } = mockedGenerateClient([
      {
        data: {
          getBlog: blogWithSelectionSet,
        },
      },
    ]);

    // simulated amplifyconfiguration.json
    const config = await buildAmplifyConfig(schema);
    // #endregion mocking

    // #region covers c5a9d7a5cdc10664
    Amplify.configure(config);

    const client = generateClient<Schema>();

    type T = Schema['Blog']['__meta']['flatModel'];

    // same way for all CRUDL: .create, .get, .update, .delete, .list, .observeQuery
    const { data: blogWithSubsetOfData, errors } = await client.models.Blog.get(
      { id: '<MY_BLOG_ID>' },
      {
        selectionSet: [
          'author.email',
          'publication.company.location.city',
          'content.*',
        ],
      },
    );
    // #endregion

    // #region assertions
    expect(optionsAndHeaders(spy)).toMatchSnapshot();
    expect(errors).toBeUndefined();
    expect(blogWithSubsetOfData).toEqual(blogWithSelectionSetResult);
    // #endregion assertions
  });

  test('TypeScript type helpers for Amplify Data part 1', async () => {
    // #region mocking
    // This example uses a more complicated schema to demonstrate CSS
    const schema = a
      .schema({
        Post: a.model({
          content: a.string(),
          author: a.string(),
          comments: a.hasMany('Comment', 'postId'),
        }),
        Comment: a.model({
          content: a.string(),
          author: a.string(),
          postId: a.id(),
          post: a.belongsTo('Post', 'postId'),
        }),
      })
      .authorization((allow) => allow.publicApiKey());
    type Schema = ClientSchema<typeof schema>;

    // #endregion mocking

    // #region covers 4c106be9c8e3749e
    type Post = Schema['Post']['type'];

    const [post, setPosts] = useState<Post[]>([]);
    // #endregion

    // #region assertions
    // not shown in the docs snippet, but setPosts should be callable with a list of post
    const newPost: Post = {
      id: 'some-id',
      createdAt: '123',
      updatedAt: '455',

      // normally will be a lazy loader as returned a `.models` method
      comments: {} as any,
    };
    expect(() => setPosts([newPost])).not.toThrow();
    // #endregion assertions
  });

  test('TypeScript type helpers for Amplify Data part 2', async () => {
    // #region mocking
    // This example uses a more complicated schema to demonstrate CSS
    const schema = a
      .schema({
        Post: a.model({
          content: a.string(),
          author: a.string(),
          comments: a.hasMany('Comment', 'postId'),
        }),
        Comment: a.model({
          content: a.string(),
          author: a.string(),
          postId: a.id(),
          post: a.belongsTo('Post', 'postId'),
        }),
      })
      .authorization((allow) => allow.publicApiKey());
    type Schema = ClientSchema<typeof schema>;

    // simulated amplifyconfiguration.json
    const config = await buildAmplifyConfig(schema);
    Amplify.configure(config);

    const { spy, generateClient } = mockedGenerateClient([
      {
        data: {
          listPosts: [],
        },
      },
    ]);
    const client = generateClient<Schema>();
    // #endregion mocking

    // #region covers 074ff31ac8fcd4d4, ab5b1a4b0db8b096, 594df1fdc5fc0b68
    const selectionSet = ['content', 'author', 'comments.*'] as const;
    type PostWithComments = SelectionSet<
      Schema['Post']['type'],
      typeof selectionSet
    >;

    const [posts, setPosts] = useState<PostWithComments[]>([]);

    async function fetchPosts() {
      const { data: postsWithComments } = await client.models.Post.list({
        selectionSet,
      });

      setPosts(postsWithComments);
    }
    // #endregion

    // #region assertions
    expect(fetchPosts).not.toThrow();
    expect(optionsAndHeaders(spy)).toMatchSnapshot();
    // #endregion assertions
  });

  // TODO: Cancel read requests
  // https://docs.amplify.aws/gen2/build-a-backend/data/query-data/#cancel-read-requests
});
