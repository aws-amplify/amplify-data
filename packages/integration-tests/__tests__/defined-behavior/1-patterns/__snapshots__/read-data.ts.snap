// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Read application data Compound filters 1`] = `
[
  [
    {
      "apiKey": undefined,
      "authMode": undefined,
      "authToken": undefined,
      "endpoint": undefined,
      "query": "query($filter: ModelTodoFilterInput,$limit: Int,$nextToken: String) { listTodos(filter: $filter,limit: $limit,nextToken: $nextToken) { items { id content description done priority createdAt updatedAt owner } nextToken __typename } }",
      "variables": {
        "filter": {
          "or": [
            {
              "priority": {
                "eq": 1,
              },
            },
            {
              "priority": {
                "eq": 2,
              },
            },
          ],
        },
      },
    },
    {},
  ],
]
`;

exports[`Read application data Fetch only the data you need with custom selection set 1`] = `
[
  [
    {
      "apiKey": undefined,
      "authMode": undefined,
      "authToken": undefined,
      "endpoint": undefined,
      "query": "query($id: ID!) { getBlog(id: $id) { author { email } publication { company { location { city } } } content { items { id title description blogId createdAt updatedAt } } } }",
      "variables": {
        "id": "<MY_BLOG_ID>",
      },
    },
    {},
  ],
]
`;

exports[`Read application data Filter list queries 1`] = `
[
  [
    {
      "apiKey": undefined,
      "authMode": undefined,
      "authToken": undefined,
      "endpoint": undefined,
      "query": "query($filter: ModelTodoFilterInput,$limit: Int,$nextToken: String) { listTodos(filter: $filter,limit: $limit,nextToken: $nextToken) { items { id content description done priority createdAt updatedAt owner } nextToken __typename } }",
      "variables": {
        "filter": {
          "content": {
            "beginsWith": "hello",
          },
        },
      },
    },
    {},
  ],
]
`;

exports[`Read application data List and get your data 1`] = `
[
  [
    {
      "apiKey": undefined,
      "authMode": undefined,
      "authToken": undefined,
      "endpoint": undefined,
      "query": "query($filter: ModelTodoFilterInput,$limit: Int,$nextToken: String) { listTodos(filter: $filter,limit: $limit,nextToken: $nextToken) { items { id content description done priority createdAt updatedAt owner } nextToken __typename } }",
      "variables": {},
    },
    {},
  ],
  [
    {
      "apiKey": undefined,
      "authMode": undefined,
      "authToken": undefined,
      "endpoint": undefined,
      "query": "query($productId: ID,$sortDirection: ModelSortDirection,$warehouseId: ModelIDKeyConditionInput,$filter: ModelInventoryFilterInput,$limit: Int,$nextToken: String) { listInventories(productId: $productId,sortDirection: $sortDirection,warehouseId: $warehouseId,filter: $filter,limit: $limit,nextToken: $nextToken) { items { productId warehouseId description createdAt updatedAt owner } nextToken __typename } }",
      "variables": {
        "productId": "...",
        "warehouseId": {
          "beginsWith": "...",
        },
      },
    },
    {},
  ],
  [
    {
      "apiKey": undefined,
      "authMode": undefined,
      "authToken": undefined,
      "endpoint": undefined,
      "query": "query($productId: ID!,$warehouseId: ID!) { getInventory(productId: $productId,warehouseId: $warehouseId) { productId warehouseId description createdAt updatedAt owner } }",
      "variables": {
        "productId": "...",
        "warehouseId": "...",
      },
    },
    {},
  ],
]
`;

exports[`Read application data Paginate list queries 1`] = `
[
  [
    {
      "apiKey": undefined,
      "authMode": undefined,
      "authToken": undefined,
      "endpoint": undefined,
      "query": "query($filter: ModelTodoFilterInput,$limit: Int,$nextToken: String) { listTodos(filter: $filter,limit: $limit,nextToken: $nextToken) { items { id content description done priority createdAt updatedAt owner } nextToken __typename } }",
      "variables": {
        "limit": 100,
        "nextToken": "eyJ2ZXJzaW9uejE1a2...",
      },
    },
    {},
  ],
]
`;

exports[`Read application data TypeScript type helpers for Amplify Data part 2 1`] = `
[
  [
    {
      "apiKey": undefined,
      "authMode": undefined,
      "authToken": undefined,
      "endpoint": undefined,
      "query": "query($filter: ModelPostFilterInput,$limit: Int,$nextToken: String) { listPosts(filter: $filter,limit: $limit,nextToken: $nextToken) { items { content author comments { items { id content author postId createdAt updatedAt } } } nextToken __typename } }",
      "variables": {},
    },
    {},
  ],
]
`;
